README.md
[![Build Status](https://travis-ci.org/onesky/api-library-php5.svg)](https://travis-ci.org/onesky/api-library-php5)
[![Code Coverage](https://scrutinizer-ci.com/g/onesky/api-library-php5/badges/coverage.png?b=master)](https://scrutinizer-ci.com/g/onesky/api-library-php5/?branch=master)

# Onesky PHP5 API Library

PHP5 library for OneSky API

## Requirements
- PHP 5.1 or higher
- CURL extension

## Installation
Install the latest version with `composer require onesky/api-library-php5`

## How to use

**Create instance**

```php
use Onesky\Api\Client;

$client = new Client();
```

**Set API key and secret**

```php
$client->setApiKey('<api-key>')->setSecret('<api-secret>');
```

**Way to make request**

```php
// resource   => name of resource in camelcase with 's' tailing such as 'projectTypes', 'quotations', 'importTasks'
// action     => action to take such as 'list', 'show', 'upload', 'export'
// parameters => parameters passed in the request including URL param such as 'project_id', 'files', 'locale'
$client->{resource}('{action}', array({parameters}));
```

**Sample request and get response in array**

```php5
// list project groups
$response = $client->projectTypes('list');
$response = json_decode($response, true);

// show a project
$response = $client->projects('show', array('project_id' => 999));
$response = json_decode($response, true);

// upload file
$response = $client->files('upload', array(
    'project_id'  => 999,
    'file'        => 'path/to/string.yml',
    'file_format' => 'YAML',
    'locale'      => 'fr'
));
$response = json_decode($response, true);

// export translation
$response = $client->translations('export', array(
    'project_id' => 999,
    'locale'     => 'ja',
    'source_file_name' => 'string.yml'
));
file_put_contents('path/to/file', $response);

// create order
$response = $client->orders('create', array(
    'project_id' => 999,
    'files'      => 'string.yml',
    'to_locale'  => 'de'
));
$response = json_decode($response, true);
```

## How to contribute

* [Raise issues or questions](https://github.com/onesky/api-library-php5/issues)
* [Create Pull Requests](https://github.com/onesky/api-library-php5/pulls)

Or pick one of [the issues](https://github.com/onesky/api-library-php5/issues) and create a PR to solve it.

## How to run tests

Note: need to install dependencies using [composer](https://getcomposer.org/) to have phpunit.

```sh
vendor/bin/phpunit
```


CHANGELOG
CHANGELOG
=========

0.1.0 (2014-01-22)
- start project

0.1.1 (2014-07-17)
- add params to request body for DELETE method

1.0.0 (2014-12-11)
- composer integration (special thanks to @k3nn7)
  - impose namespace
  - add license

1.0.1 (2014-12-11)
- fix include parameter with boolean value in request body

1.0.2 (2014-12-29)
- fix deprecated warning of using @ prefix to upload file

1.0.3 (2014-12-30)
- add custom http header

1.0.4 (2015-02-09)
- add list import tasks (special thanks to @brandesign)
- specify namespace

1.0.5 (2015-06-09)
- remove hardcoded version (thanks to @skolodyazhnyy)
- apply PSR-4 and revise coding style (thanks to @BunnyHolder)
- add file format checking (thanks to @BunnyHolder)
- setup phpunit (thanks to @BernardoSilva)
- activate TravisCI

src/Onesky/Api/Tests/ApiTest.php
<?php

namespace Onesky\Api\Tests;

use Onesky\Api\Client;

class ApiTest extends \PHPUnit_Framework_TestCase
{
    /** @var \Onesky\Api\Client $api */
    protected $api;

    public function setUp()
    {
        $this->api = new Client();
    }

    public function testThatApiClassInstantiatesProperly()
    {
        $this->assertInstanceOf('Onesky\Api\Client', $this->api);
    }
}


src/Onesky/Api/Tests/FilesApiTest.php
<?php

namespace Onesky\Api\Tests;

use Onesky\Api\Client;

/**
 * Class to test the files resource.
 *
 * See more about files resource here:
 * https://github.com/onesky/api-documentation-platform/blob/master/resources/file.md
 *
 * @author Bernardo Silva <benny.stuff@gmail.com>
 */
class FilesApiTest extends \PHPUnit_Framework_TestCase
{
    /** @var \Onesky\Api\Client $api */
    protected $api;

    public function setUp()
    {
        $this->api = new Client();
    }

    /**
     * @expectedException \InvalidArgumentException
     * @expectedExceptionMessage Invalid resource action
     */
    public function testFilesDetectMissingResourceAction()
    {
        $this->api->files();
    }

    /**
     * @expectedException \InvalidArgumentException
     * @expectedExceptionMessage Missing parameter: project_id
     */
    public function testFilesListDetectMissingParam()
    {
        $this->api->files('list');
    }

    /**
     * @expectedException \UnexpectedValueException
     * @expectedExceptionMessage Invalid authenticate data of api key or secret
     */
    public function testFilesListDetectInvalidAuthentication()
    {
        $this->api->files('list', ['project_id' => 123]);
    }


    /**
     * @expectedException \InvalidArgumentException
     * @expectedExceptionMessage Missing parameter: project_id
     */
    public function testFilesUploadDetectMissingParam()
    {
        $this->api->files('upload');
    }

    /**
     * @expectedException \UnexpectedValueException
     * @expectedExceptionMessage Invalid authenticate data of api key or secret
     */
    public function testFilesUploadDetectInvalidAuthentication()
    {
        $this->api->files('upload', ['project_id' => 123]);
    }

    /**
     * @expectedException \InvalidArgumentException
     * @expectedExceptionMessage Missing parameter: project_id
     */
    public function testFilesDeleteDetectMissingParam()
    {
        $this->api->files('delete');
    }

    /**
     * @expectedException \UnexpectedValueException
     * @expectedExceptionMessage Invalid authenticate data of api key or secret
     */
    public function testFilesDeleteDetectInvalidAuthentication()
    {
        $this->api->files('delete', ['project_id' => 123]);
    }
}


src/Onesky/Api/Tests/ProjectsApiTest.php
<?php

namespace Onesky\Api\Tests;

use Onesky\Api\Client;

/**
 * Class to test the projects resource.
 *
 * See more about projects resource here:
 * https://github.com/onesky/api-documentation-platform/blob/master/resources/project.md
 *
 * @author Bernardo Silva <benny.stuff@gmail.com>
 */
class ProjectsApiTest extends \PHPUnit_Framework_TestCase
{
    /** @var \Onesky\Api\Client $api */
    protected $api;

    public function setUp()
    {
        $this->api = new Client();
    }

    /**
     * @expectedException \InvalidArgumentException
     * @expectedExceptionMessage Invalid resource action
     */
    public function testProjectsDetectMissingResourceAction()
    {
        $this->api->projects();
    }

    /**
     * @expectedException \InvalidArgumentException
     * @expectedExceptionMessage Missing parameter: project_group_id
     */
    public function testProjectsListDetectMissingGroupIdParam()
    {
        $this->api->projects('list');
    }

    /**
     * @expectedException \UnexpectedValueException
     * @expectedExceptionMessage Invalid authenticate data of api key or secret
     */
    public function testProjectsListDetectInvalidAuthentication()
    {
        $this->api->projects('list', ['project_group_id' => 123]);
    }


    /**
     * @expectedException \InvalidArgumentException
     * @expectedExceptionMessage Missing parameter: project_id
     */
    public function testProjectsShowDetectMissingParam()
    {
        $this->api->projects('show');
    }

    /**
     * @expectedException \UnexpectedValueException
     * @expectedExceptionMessage Invalid authenticate data of api key or secret
     */
    public function testProjectsShowDetectInvalidAuthentication()
    {
        $this->api->projects('show', ['project_id' => 123]);
    }

    /**
     * @expectedException \InvalidArgumentException
     * @expectedExceptionMessage Missing parameter: project_group_id
     */
    public function testProjectsCreateDetectMissingGroupIdParam()
    {
        $this->api->projects('create');
    }

    /**
     * @expectedException \UnexpectedValueException
     * @expectedExceptionMessage Invalid authenticate data of api key or secret
     */
    public function testProjectsCreateDetectMissingParam()
    {
        $this->api->projects('create', ['project_group_id' => 123]);
    }

    /**
     * @expectedException \InvalidArgumentException
     * @expectedExceptionMessage Missing parameter: project_id
     */
    public function testProjectsUpdateDetectMissingParam()
    {
        $this->api->projects('update');
    }

    /**
     * @expectedException \UnexpectedValueException
     * @expectedExceptionMessage Invalid authenticate data of api key or secret
     */
    public function testProjectsUpdateDetectInvalidAuthentication()
    {
        $this->api->projects('update', ['project_id' => 123]);
    }

    /**
     * @expectedException \InvalidArgumentException
     * @expectedExceptionMessage Missing parameter: project_id
     */
    public function testProjectsDeleteDetectMissingParam()
    {
        $this->api->projects('delete');
    }

    /**
     * @expectedException \UnexpectedValueException
     * @expectedExceptionMessage Invalid authenticate data of api key or secret
     */
    public function testProjectsDeleteDetectInvalidAuthentication()
    {
        $this->api->projects('delete', ['project_id' => 123]);
    }

    /**
     * @expectedException \InvalidArgumentException
     * @expectedExceptionMessage Missing parameter: project_id
     */
    public function testProjectsLanguagesDetectMissingParam()
    {
        $this->api->projects('languages');
    }

    /**
     * @expectedException \UnexpectedValueException
     * @expectedExceptionMessage Invalid authenticate data of api key or secret
     */
    public function testProjectsLanguagesDetectInvalidAuthentication()
    {
        $this->api->projects('languages', ['project_id' => 123]);
    }
}


src/api-library-php5.php
<?php

/**
 * Onesky API wrapper PHP5 library
 */

if (!function_exists('curl_init')) {
  throw new Exception('OneSky needs the CURL PHP extension.');
}

class Onesky_Api
{
    /**
     * Onesky API endpoint
     */
    protected $_endpoint = 'https://platform.api.onesky.io/1';

    /**
     * Client authenticate token
     */
    protected $_apiKey = '';

    /**
     * Client authenticate secret
     */
    protected $_secret = '';

    /**
     * Resources with actions
     */
    protected $_resources = array(
        'project_groups' => array(
            'list'      => '/project-groups',
            'show'      => '/project-groups/:project_group_id',
            'create'    => '/project-groups',
            'delete'    => '/project-groups/:project_group_id',
            'languages' => '/project-groups/:project_group_id/languages',
        ),
        'projects'       => array(
            'list'      => '/project-groups/:project_group_id/projects',
            'show'      => '/projects/:project_id',
            'create'    => '/project-groups/:project_group_id/projects',
            'update'    => '/projects/:project_id',
            'delete'    => '/projects/:project_id',
            'languages' => '/projects/:project_id/languages',
        ),
        'files'          => array(
            'list'   => '/projects/:project_id/files',
            'upload' => '/projects/:project_id/files',
            'delete' => '/projects/:project_id/files',
        ),
        'translations'   => array(
            'export' => '/projects/:project_id/translations',
            'status' => '/projects/:project_id/translations/status',
        ),
        'import_tasks'   => array(
            'show' => '/projects/:project_id/import-tasks/:import_id'
        ),
        'quotations'     => array(
            'show' => '/projects/:project_id/quotations'
        ),
        'orders'         => array(
            'list'   => '/projects/:project_id/orders',
            'show'   => '/projects/:project_id/orders/:order_id',
            'create' => '/projects/:project_id/orders'
        ),
        'locales'        => array(
            'list' => '/locales'
        ),
        'project_types'  => array(
            'list' => '/project-types'
        ),
    );

    /**
     * Actions to use multipart to upload file
     */
    protected $_multiPartActions = array(
        'files' => array('upload'),
    );

    /**
     * Actions to use multipart to upload file
     */
    protected $_exportFileActions = array(
        'translations' => array('export'),
    );

    /**
     * Methods of actions mapping
     */
    protected $_methods = array(
        // 'get'    => array('list', 'show', 'languages', 'export', 'status'),
        'post'   => array('create', 'upload'),
        'put'    => array('update'),
        'delete' => array('delete'),
    );

    /**
     * Default curl settings
     */
    protected $_curlSettings = array(
        CURLOPT_RETURNTRANSFER => true,
    );

    public function setApiKey($apiKey)
    {
        $this->_apiKey = $apiKey;
        return $this;
    }

    public function setSecret($secret)
    {
        $this->_secret = $secret;
        return $this;
    }

    /**
     * Retrieve resources
     * @return array
     */
    public function getResources()
    {
        return array_keys($this->_resources);
    }

    /**
     * Retrieve actions of a resource
     * @param  string $resource Resource name
     * @return array|null
     */
    public function getActionsByResource($resource)
    {
        if (!isset($this->_resources[$resource]))
            return null; // no resource found

        $actions = array();
        foreach ($this->_resources[$resource] as $action => $path)
            $actions[] = $action;

        return $actions;
    }

    /**
     * Retrieve the corresponding method to use
     * @param  string $action Action name
     * @return string
     */
    public function getMethodByAction($action)
    {
        foreach ($this->_methods as $method => $actions) {
            if (in_array($action, $actions))
                return $method;
        }

        return 'get';
    }

    /**
     * Determine if using mulit-part to upload file
     * @param  string  $resource Resource name
     * @param  string  $action   Action name
     * @return boolean
     */
    public function isMultiPartAction($resource, $action)
    {
        return isset($this->_multiPartActions[$resource]) && in_array($action, $this->_multiPartActions[$resource]);
    }

    /**
     * Determine if it is to export (download) file
     * @param  string  $resource Resource name
     * @param  string  $action   Action name
     * @return boolean
     */
    public function isExportFileAction($resource, $action)
    {
        return isset($this->_exportFileActions[$resource]) && in_array($action, $this->_exportFileActions[$resource]);
    }

    /**
     * For developers to initial request to Onesky API
     * 
     * Example:
     *     $onesky = new Onesky_Api();
     *     $onesky->setApiKey('<api-key>')->setSecret('<api-secret>');
     *     
     *     // To list project type
     *     $onesky->projectTypes('list');
     *
     *     // To create project
     *     $onesky->projects('create', array('project_group_id' => 999));
     *
     *     // To upload string file
     *     $onesky->files('upload', array('project_id' => 1099, 'file' => 'path/to/file.yml', 'file_format' => 'YAML'));
     * 
     * @param  string $fn_name Function name acted as resource name
     * @param  array  $params  Parameters passed in request
     * @return array  Response
     */
    public function __call($fn_name, $params)
    {
        // is valid resource
        $resource = strtolower(preg_replace('/([a-z])([A-Z])/', '$1_$2', $fn_name)); // camelcase to underscore
        if (!in_array($resource, $this->getResources()))
            throw new BadMethodCallException('Invalid resource');

        // is valid action
        $action = array_shift($params); // action name
        if (!in_array($action, $this->getActionsByResource($resource)))
            throw new InvalidArgumentException('Invalid resource action');

        // parameters
        if (count($params) > 0) {
            $params = $this->_normalizeParams(array_shift($params));
        } else {
            $params = array();
        }

        // get request method
        $method = $this->getMethodByAction($action);

        // get path
        $path = $this->_getRequestPath($resource, $action, $params);

        // is multi-part
        $isMultiPart = $this->isMultiPartAction($resource, $action);

        // return response
        return $this->_callApi($method, $path, $params, $isMultiPart);
    }

    /**
     * Retrieve request path and replace variables with values
     * @param  string $resource Resource name
     * @param  string $action   Action name
     * @param  array  $params   Parameters
     * @return string Request path
     */
    private function _getRequestPath($resource, $action, &$params)
    {
        if (!isset($this->_resources[$resource]) || !isset($this->_resources[$resource][$action]))
            throw new UnexpectedValueException('Resource path not found');

        // get path
        $path = $this->_resources[$resource][$action];

        // replace variables
        $matchCount = preg_match_all("/:(\w*)/", $path, $variables);
        if ($matchCount) {
            foreach ($variables[0] as $index => $placeholder) {
                if (!isset($params[$variables[1][$index]]))
                    throw new InvalidArgumentException('Missing parameter: ' . $variables[1][$index]);

                $path = str_replace($placeholder, $params[$variables[1][$index]], $path);
                unset($params[$variables[1][$index]]); // remove parameter from $params
            }
        }

        return $path;
    }

    protected function _verifyTokenAndSecret()
    {
        if (empty($this->_apiKey) || empty($this->_secret))
            throw new UnexpectedValueException('Invalid authenticate data of api key or secret');
    }

    /**
     * Initial request to Onesky API
     * @param  string  $method
     * @param  string  $path
     * @param  array   $params
     * @param  boolean $isMultiPart
     * @param  boolean $isExportFile
     * @return array
     */
    private function _callApi($method, $path, $params, $isMultiPart)
    {
        // init session
        $ch = curl_init();

        // request settings
        curl_setopt_array($ch, $this->_curlSettings); // basic settings

        // url
        $url = $this->_endpoint . $path;
        if ($method == 'get') // ['post', 'put', 'delete']
            $url .= $this->_getAuthQueryStringWithParams($params);
        else
            $url .= $this->_getAuthQueryString();
        curl_setopt($ch, CURLOPT_URL, $url);

        // http header
        if (!$isMultiPart)
            curl_setopt($ch, CURLOPT_HTTPHEADER, array("Content-Type: application/json"));

        // method specific settings
        switch ($method) {
            case 'post':
                curl_setopt($ch, CURLOPT_POST, 1);

                // requst body
                if ($isMultiPart) {
                    $params['file'] = '@' . $params['file'];
                    $postBody = $params;
                } else {
                    $postBody = json_encode($params);
                }
                curl_setopt($ch, CURLOPT_POSTFIELDS, $postBody);

                break;

            case 'put':
                curl_setopt($ch, CURLOPT_CUSTOMREQUEST, "PUT");
                curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($params));

                break;

            case 'delete':
                curl_setopt($ch, CURLOPT_CUSTOMREQUEST, "DELETE");
                curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($params));

                break;
        }

        // execute request
        $response = curl_exec($ch);

        // error handling
        if ($response === false)
            throw new UnexpectedValueException(curl_error($ch));

        // close connection
        curl_close($ch);

        // return response
        return $response;
    }

    private function _getAuthQueryStringWithParams($params)
    {
        $queryString = $this->_getAuthQueryString();

        if (count($params) > 0)
            $queryString .= '&' . http_build_query($params);

        return $queryString;
    }

    private function _getAuthQueryString()
    {
        $this->_verifyTokenAndSecret();

        $timestamp = time();
        $devHash = md5($timestamp . $this->_secret);

        $queryString  = '?api_key=' . $this->_apiKey;
        $queryString .= '&timestamp=' . $timestamp;
        $queryString .= '&dev_hash=' . $devHash;

        return $queryString;
    }

    private function _normalizeParams(array $params)
    {
        // change boolean value to integer for curl
        foreach ($params as $key => $value) {
            if (is_bool($value)) {
                $params[$key] = (int)$value;
            }
        }

        return $params;
    }
}


src/Onesky/Api/FileFormat.php
<?php

namespace Onesky\Api;

/**
 * Class FileType
 */
class FileFormat
{
    const IOS_STRINGS = 'IOS_STRINGS';
    const IOS_STRINGSDICT_XML = 'IOS_STRINGSDICT_XML';
    const GNU_PO = 'GNU_PO';
    const ANDROID_XML = 'ANDROID_XML';
    const ANDROID_JSON = 'ANDROID_JSON';
    const JAVA_PROPERTIES = 'JAVA_PROPERTIES';
    const RUBY_YML = 'RUBY_YML';
    const RUBY_YAML = 'RUBY_YAML';
    const FLASH_XML = 'FLASH_XML';
    const GNU_POT = 'GNU_POT';
    const RRC = 'RRC';
    const RESX = 'RESX';
    const RESJSON = 'RESJSON';
    const HIERARCHICAL_JSON = 'HIERARCHICAL_JSON';
    const PHP = 'PHP';
    const PHP_VARIABLES = 'PHP_VARIABLES';
    const HTML = 'HTML';
    const RESW = 'RESW';
    const YML = 'YML';
    const YAML = 'YAML';
    const ADEMPIERE_XML = 'ADEMPIERE_XML';
    const IDEMPIERE_XML = 'IDEMPIERE_XML';
    const QT_TS_XML = 'QT_TS_XML';
    const TMX = 'TMX';
    const L10N = 'L10N';
    const INI = 'INI';
    const REQUIREJS = 'REQUIREJS';

    /**
     * @var array
     */
    protected $supportedFormats = array(
        self::IOS_STRINGS,
        self::IOS_STRINGSDICT_XML,
        self::GNU_PO,
        self::ANDROID_XML,
        self::ANDROID_JSON,
        self::JAVA_PROPERTIES,
        self::RUBY_YML,
        self::RUBY_YAML,
        self::FLASH_XML,
        self::GNU_POT,
        self::RRC,
        self::RESX,
        self::RESJSON,
        self::HIERARCHICAL_JSON,
        self::PHP,
        self::PHP_VARIABLES,
        self::HTML,
        self::RESW,
        self::YML,
        self::YAML,
        self::ADEMPIERE_XML,
        self::IDEMPIERE_XML,
        self::QT_TS_XML,
        self::TMX,
        self::L10N,
        self::INI,
        self::REQUIREJS
    );
}

src/Onesky/Api/Client.php
<?php

namespace Onesky\Api;

/**
 * Onesky API wrapper PHP5 library
 */
class Client
{
    /**
     * Onesky API endpoint
     */
    protected $endpoint = 'https://platform.api.onesky.io/1';

    /**
     * Client authenticate token
     */
    protected $apiKey = null;

    /**
     * Client authenticate secret
     */
    protected $secret = null;

    /**
     * Resources with actions
     */
    protected $resources = array(
        'project_groups' => array(
            'list'      => '/project-groups',
            'show'      => '/project-groups/:project_group_id',
            'create'    => '/project-groups',
            'delete'    => '/project-groups/:project_group_id',
            'languages' => '/project-groups/:project_group_id/languages',
        ),
        'projects'       => array(
            'list'      => '/project-groups/:project_group_id/projects',
            'show'      => '/projects/:project_id',
            'create'    => '/project-groups/:project_group_id/projects',
            'update'    => '/projects/:project_id',
            'delete'    => '/projects/:project_id',
            'languages' => '/projects/:project_id/languages',
        ),
        'files'          => array(
            'list'   => '/projects/:project_id/files',
            'upload' => '/projects/:project_id/files',
            'delete' => '/projects/:project_id/files',
        ),
        'translations'   => array(
            'export' => '/projects/:project_id/translations',
            'status' => '/projects/:project_id/translations/status',
        ),
        'import_tasks'   => array(
	    'list' => '/projects/:project_id/import-tasks/',
            'show' => '/projects/:project_id/import-tasks/:import_id'
        ),
        'quotations'     => array(
            'show' => '/projects/:project_id/quotations'
        ),
        'orders'         => array(
            'list'   => '/projects/:project_id/orders',
            'show'   => '/projects/:project_id/orders/:order_id',
            'create' => '/projects/:project_id/orders'
        ),
        'locales'        => array(
            'list' => '/locales'
        ),
        'project_types'  => array(
            'list' => '/project-types'
        ),
    );

    /**
     * Actions to use multipart to upload file
     */
    protected $multiPartActions = array(
        'files' => array('upload'),
    );

    /**
     * Actions to use multipart to upload file
     */
    protected $exportFileActions = array(
        'translations' => array('export'),
    );

    /**
     * Methods of actions mapping
     */
    protected $methods = array(
        // 'get'    => array('list', 'show', 'languages', 'export', 'status'),
        'post'   => array('create', 'upload'),
        'put'    => array('update'),
        'delete' => array('delete'),
    );

    /**
     * Default curl settings
     */
    protected $curlSettings = array(
        CURLOPT_RETURNTRANSFER => true,
    );

    protected $httpHeaders = array(
        "Onesky-Plugin: php-wrapper",
    );

    public function __construct()
    {
        if (!function_exists('curl_init')) {
            throw new \Exception('OneSky needs the CURL PHP extension.');
        }
    }

    /**
     * @param string $apiKey
     * @return $this
     */
    public function setApiKey($apiKey)
    {
        $this->apiKey = $apiKey;
        return $this;
    }

    /**
     * @param string $secret
     * @return $this
     */
    public function setSecret($secret)
    {
        $this->secret = $secret;
        return $this;
    }

    /**
     * Retrieve resources
     * @return array
     */
    public function getResources()
    {
        return array_keys($this->resources);
    }

    /**
     * Retrieve actions of a resource
     * @param  string $resource Resource name
     * @return array|null
     */
    public function getActionsByResource($resource)
    {
        if (!isset($this->resources[$resource])) {
            return null; // no resource found
        }

        $actions = array();
        foreach ($this->resources[$resource] as $action => $path) {
            $actions[] = $action;
        }

        return $actions;
    }

    /**
     * Retrieve the corresponding method to use
     * @param  string $action Action name
     * @return string
     */
    public function getMethodByAction($action)
    {
        foreach ($this->methods as $method => $actions) {
            if (in_array($action, $actions)) {
                return $method;
            }
        }

        return 'get';
    }

    /**
     * Determine if using mulit-part to upload file
     * @param  string  $resource Resource name
     * @param  string  $action   Action name
     * @return boolean
     */
    public function isMultiPartAction($resource, $action)
    {
        return isset($this->multiPartActions[$resource]) && in_array($action, $this->multiPartActions[$resource]);
    }

    /**
     * Determine if it is to export (download) file
     * @param  string  $resource Resource name
     * @param  string  $action   Action name
     * @return boolean
     */
    public function isExportFileAction($resource, $action)
    {
        return isset($this->exportFileActions[$resource]) && in_array($action, $this->exportFileActions[$resource]);
    }

    /**
     * For developers to initial request to Onesky API
     * 
     * Example:
     *     $onesky = new Onesky_Api();
     *     $onesky->setApiKey('<api-key>')->setSecret('<api-secret>');
     *     
     *     // To list project type
     *     $onesky->projectTypes('list');
     *
     *     // To create project
     *     $onesky->projects('create', array('project_group_id' => 999));
     *
     *     // To upload string file
     *     $onesky->files('upload', array('project_id' => 1099, 'file' => 'path/to/file.yml', 'file_format' => 'YAML'));
     * 
     * @param  string $fn_name Function name acted as resource name
     * @param  array  $params  Parameters passed in request
     * @return array  Response
     */
    public function __call($fn_name, $params)
    {
        // is valid resource
        $resource = strtolower(preg_replace('/([a-z])([A-Z])/', '$1_$2', $fn_name)); // camelcase to underscore
        if (!in_array($resource, $this->getResources())) {
            throw new \BadMethodCallException('Invalid resource');
        }

        // is valid action
        $action = array_shift($params); // action name
        if (!in_array($action, $this->getActionsByResource($resource))) {
            throw new \InvalidArgumentException('Invalid resource action');
        }

        // parameters
        if (count($params) > 0) {
            $params = $this->normalizeParams(array_shift($params));
        } else {
            $params = array();
        }

        // get request method
        $method = $this->getMethodByAction($action);

        // get path
        $path = $this->getRequestPath($resource, $action, $params);

        // is multi-part
        $isMultiPart = $this->isMultiPartAction($resource, $action);

        // return response
        return $this->callApi($method, $path, $params, $isMultiPart);
    }

    /**
     * Retrieve request path and replace variables with values
     * @param  string $resource Resource name
     * @param  string $action   Action name
     * @param  array  $params   Parameters
     * @return string Request path
     */
    private function getRequestPath($resource, $action, &$params)
    {
        if (!isset($this->resources[$resource]) || !isset($this->resources[$resource][$action])) {
            throw new \UnexpectedValueException('Resource path not found');
        }

        // get path
        $path = $this->resources[$resource][$action];

        // replace variables
        $matchCount = preg_match_all("/:(\w*)/", $path, $variables);
        if ($matchCount) {
            foreach ($variables[0] as $index => $placeholder) {
                if (!isset($params[$variables[1][$index]])) {
                    throw new \InvalidArgumentException('Missing parameter: ' . $variables[1][$index]);
                }

                $path = str_replace($placeholder, $params[$variables[1][$index]], $path);
                unset($params[$variables[1][$index]]); // remove parameter from $params
            }
        }

        return $path;
    }

    protected function verifyTokenAndSecret()
    {
        if (empty($this->apiKey) || empty($this->secret)) {
            throw new \UnexpectedValueException('Invalid authenticate data of api key or secret');
        }
    }

    /**
     * Initial request to Onesky API
     * @param  string  $method
     * @param  string  $path
     * @param  array   $params
     * @param  boolean $isMultiPart
     * @return array
     */
    private function callApi($method, $path, $params, $isMultiPart)
    {
        // init session
        $ch = curl_init();

        // request settings
        curl_setopt_array($ch, $this->curlSettings); // basic settings

        // url
        $url = $this->endpoint . $path;
        $url .= $method == 'get' ? $this->getAuthQueryStringWithParams($params) : $this->getAuthQueryString();

        curl_setopt($ch, CURLOPT_URL, $url);

        // http header
        $requestHeaders = $this->httpHeaders;
        if (!$isMultiPart) {
            $requestHeaders[] = "Content-Type: application/json";
        }
        curl_setopt($ch, CURLOPT_HTTPHEADER, $requestHeaders);

        // method specific settings
        switch ($method) {
            case 'post':
                curl_setopt($ch, CURLOPT_POST, 1);

                // request body
                if ($isMultiPart) {
                    if (version_compare(PHP_VERSION, '5.5.0') === -1) {
                        // fallback to old method
                        $params['file'] = '@' . $params['file'];
                    } else {
                        // make use of CURLFile
                        curl_setopt($ch, CURLOPT_SAFE_UPLOAD, true);
                        $params['file'] = new \CURLFile($params['file']);
                    }
                    $postBody = $params;
                } else {
                    $postBody = json_encode($params);
                }
                curl_setopt($ch, CURLOPT_POSTFIELDS, $postBody);
                break;

            case 'put':
                curl_setopt($ch, CURLOPT_CUSTOMREQUEST, "PUT");
                curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($params));
                break;

            case 'delete':
                curl_setopt($ch, CURLOPT_CUSTOMREQUEST, "DELETE");
                curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($params));
                break;
        }

        // execute request
        $response = curl_exec($ch);

        // error handling
        if ($response === false) {
            throw new \UnexpectedValueException(curl_error($ch));
        }

        // close connection
        curl_close($ch);

        // return response
        return $response;
    }

    /**
     * @param $params
     * @return string
     */
    private function getAuthQueryStringWithParams($params)
    {
        $queryString = $this->getAuthQueryString();

        if (count($params) > 0) {
            $queryString .= '&' . http_build_query($params);
        }

        return $queryString;
    }

    /**
     * @return string
     */
    private function getAuthQueryString()
    {
        $this->verifyTokenAndSecret();

        $timestamp = time();
        $devHash = md5($timestamp . $this->secret);

        $queryString  = '?api_key=' . $this->apiKey;
        $queryString .= '&timestamp=' . $timestamp;
        $queryString .= '&dev_hash=' . $devHash;

        return $queryString;
    }

    /**
     * @param array $params
     * @return array
     */
    private function normalizeParams(array $params)
    {
        // change boolean value to integer for curl
        foreach ($params as $key => $value) {
            if (is_bool($value)) {
                $params[$key] = (int)$value;
            }
        }

        return $params;
    }
}
